arch: riscv: implement frame-pointer based stack unwinding

ycsin opened the pull request:
Influenced heavily by the RISCV64 stack unwinding implementation in the [Linux kernel](https://github.com/torvalds/linux/blob/master/arch/riscv/kernel/stacktrace.c).

https://github.com/zephyrproject-rtos/zephyr/blob/14149b1de14dc3c4a9ada3d22fe66c385d66fded/tests/arch/common/stack_unwind/src/main.c#L15-L39

```
$ west build -b qemu_riscv64 -p auto -t run zephyr/tests/arch/common/stack_unwind/
...
-- west build: running target run
[1/103] Preparing syscall dependency handling

[3/103] Generating include/generated/version.h
-- Zephyr version: 3.6.99 (/Users/ycsin/zephyrproject/zephyr), build: v3.6.0-2575-g14149b1de14d
[102/103] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
             RAM:       52260 B       256 MB      0.02%
        IDT_LIST:          0 GB         2 KB      0.00%
Generating files from /Users/ycsin/zephyrproject/build/zephyr/zephyr.elf for board: qemu_riscv64
[102/103] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: riscv64
*** Booting Zephyr OS build v3.6.0-2575-g14149b1de14d ***
Hello World! qemu_riscv64
1: func1
2: func2
3: func1
4: func2
5: func1
6: func2
E:      a0: 0000000000000003    t0: 0000000000000000
E:      a1: 0000000080008d1f    t1: 000000000000004c
E:      a2: 000000000000000a    t2: 0000000000000053
E:      a3: 000000008000a210    t3: 000000000000002a
E:      a4: 000000008000a210    t4: 000000000000002e
E:      a5: 0000000000000004    t5: 000000000000007f
E:      a6: 0000000000000068    t6: 0000000000000010
E:      a7: 000000000000006a
E:      ra: 00000000800003f8
E:    mepc: 0000000080000406
E: mstatus: 0000000a00021880
E: 
E: call trace:
E:       0: fp: 000000008000cac0   ra: 0000000080000442
E:       1: fp: 000000008000cae0   ra: 000000008000040e
E:       2: fp: 000000008000cb00   ra: 0000000080000442
E:       3: fp: 000000008000cb20   ra: 000000008000040e
E:       4: fp: 000000008000cb40   ra: 0000000080000442
E:       5: fp: 000000008000cb60   ra: 000000008000046e
E:       6: fp: 000000008000cb70   ra: 0000000080002382
E:       7: fp: 000000008000cb80   ra: 00000000800007fe
E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
E: Current thread: 0x8000a210 (main)
E: Halting system
qemu-system-riscv64: terminating on signal 2 from pid 2110 (<unknown process>)
ninja: build stopped: interrupted by user.
$ ../zephyr-sdk-0.16.5-1/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-addr2line -a 0000000080000442 -e build/zephyr/zephyr.elf
0x0000000080000442
/Users/ycsin/zephyrproject/zephyr/tests/arch/common/stack_unwind/src/main.c:30
$ ../zephyr-sdk-0.16.5-1/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-addr2line -a 000000008000040e -e build/zephyr/zephyr.elf
0x000000008000040e
/Users/ycsin/zephyrproject/zephyr/tests/arch/common/stack_unwind/src/main.c:24
$ ../zephyr-sdk-0.16.5-1/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-addr2line -a 000000008000046e -e build/zephyr/zephyr.elf
0x000000008000046e
/Users/ycsin/zephyrproject/zephyr/tests/arch/common/stack_unwind/src/main.c:39
$ ../zephyr-sdk-0.16.5-1/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-addr2line -a 0000000080002382 -e build/zephyr/zephyr.elf
0x0000000080002382
/Users/ycsin/zephyrproject/zephyr/kernel/include/kthread.h:176
$ ../zephyr-sdk-0.16.5-1/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-addr2line -a 00000000800007fe -e build/zephyr/zephyr.elf
0x00000000800007fe
/Users/ycsin/zephyrproject/zephyr/include/zephyr/kernel.h:593
```

cfriedt commented:
@ycsin - LGTM - I wonder if it's possible to add a check to a new or existing testsuite.

ycsin commented:
> @ycsin - LGTM - I wonder if it's possible to add a check to a new or existing testsuite.

Maybe in https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/kernel/fatal/exception ? I'll try.

ycsin commented on arch/riscv/core/fatal.c:
```diff
@@ -28,6 +28,84 @@ static const struct z_exc_handle exceptions[] = {
+static inline bool in_text_region(uintptr_t addr)
+{
+#ifndef CONFIG_SOC_ESP32C3
+	extern uintptr_t __text_region_start, __text_region_end;
```
@marekmatej is there a `__text_region_start` & `__text_region_end` equivalent for esp32c3?

ycsin replied:
probably `_text_start` & `_text_end`?

cfriedt replied:
Judging by `build/zephyr/zephyr.map` that looks like it should be the case

ycsin replied:
Created #71363 for this

carlocaione commented on arch/riscv/Kconfig:
```diff
@@ -38,6 +38,24 @@ config RISCV_ALWAYS_SWITCH_THROUGH_ECALL
 	  and most people should say n here to minimize context switching
 	  overhead.
 
+config RISCV_HAS_FRAME_POINTER
```
on ARM we have `ARM64_ENABLE_FRAME_POINTER`, so either we merge the two implementations or at least we use the same naming.

ycsin replied:
I prefer `HAS` instead of `ENABLE` :(
But I can do that if it's a must ¯\\_(ツ)_/¯

cfriedt replied:
Using `ENABLE` is fine. Naming things is hard anyway

ycsin replied:
Changed to `RISCV_ENABLE_FRAME_POINTER`

carlocaione commented on tests/arch/riscv/stack_unwind/prj.conf:
```diff
@@ -0,0 +1,11 @@
+CONFIG_TEST=y
```
Given that this is also supported by ARM64 we should have a non-arch specific test for this.

ycsin replied:
and x86 as well, that means we will need to have them printing stack traces in a similar way?

cfriedt replied:
Agreed, it is an annoyance that there is no standard way and that backtraces can be fragile based on optimizations and inlining.

But if there is a way that works for gcc and a way that works for llvm (possibly the same?), and there are a finite number of toolchains being supported, then that's already (scribbles on napkin) 99.9% of Zephyr's RISC-V users.

In any case, it isn't set in stone here.

This would be a great topic to raise with riscv.

ycsin replied:
Added a test that should work for riscv, arm & x86, the only difference is the regex, since these platform output different things

> I found this useless for RISCV given that no standard actually exists for stack unwinding and each compiler is doing something different.

Since the Linux kernel has similar implementation for RISC-V, hopefully it is _generic enough_ to work on _most_ of the RISC-V SOCs.

carlocaione reviewed (dismissed):
Besides the requested changes this is approach is REALLY prone to failure. There is no a standard way on RISCV to unwind the stack (see for example https://forums.sifive.com/t/how-to-get-backtrace-in-rtos-based-on-risc-v-64/5643), so this is currently working basically by chance. Do we really want to add this upstream?

ycsin committed:
arch: riscv: implement frame-pointer based stack unwinding

Influenced heavily by the RISCV64 stack unwinding
implementation in the Linux kernel.

`CONFIG_RISCV_EXCEPTION_STACK_TRACE` can be enabled by
configuring the following Kconfigs:

```prj.conf
CONFIG_DEBUG_INFO=y
CONFIG_EXCEPTION_STACK_TRACE=y
CONFIG_OVERRIDE_FRAME_POINTER_DEFAULT=y
CONFIG_OMIT_FRAME_POINTER=n
```

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: arch: common: add stack_unwind test

Added test for archs that support it. It triggers an exception
and verify that the fatal message has stack unwind info.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
Rebased and the changes to esp32c3 is now being worked out in #71363 instead, CI is expected to fail in this PR

ycsin commented:
Rebased on top of #71388, CI is green now.

ping @con-pax @cfriedt @carlocaione @fkokosinski @uLipe

ycsin committed:
doc: release: add notes on RISC-V's stack unwinding

Add an entry in the RISC-V arch section about the
implementation of frame-pointer based stack unwinding.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

nashif commented on arch/riscv/Kconfig:
```diff
@@ -38,6 +38,24 @@ config RISCV_ALWAYS_SWITCH_THROUGH_ECALL
 	  and most people should say n here to minimize context switching
 	  overhead.
 
+config RISCV_ENABLE_FRAME_POINTER
```
would be nice to make this a generic arch kconfig (CONFIG_FRAME_POINTER), we now have two architectures doing this

nashif commented on arch/riscv/Kconfig:
```diff
@@ -38,6 +38,24 @@ config RISCV_ALWAYS_SWITCH_THROUGH_ECALL
+	  and OMIT_FRAME_POINTER. It is automatically enabled when the frame
+	  pointer unwinding is enabled.
+
+config RISCV_EXCEPTION_STACK_TRACE
```
ditto
